types/validator_set_test.go

```diff
@@ -658,6 +651,81 @@ type testVal struct {
 	power int64
 }
 
+func permutation(valList []testVal) []testVal {
+	if len(valList) == 0 {
+		return nil
+	}
+	permList := make([]testVal, len(valList))
+	perm := cmn.RandPerm(len(valList))
+	for i, v := range perm {
+		permList[v] = valList[i]
+	}
+	return permList
+}
+
+func createNewValidatorList(testValList []testVal) []*Validator {
+	valList := make([]*Validator, 0, len(testValList))
+	for _, val := range testValList {
+		valList = append(valList, newValidator([]byte(val.name), val.power))
+	}
+	return valList
+}
+
+func createNewValidatorSet(testValList []testVal) *ValidatorSet {
+	valList := createNewValidatorList(testValList)
+	valSet := NewValidatorSet(valList)
+	return valSet
+}
+
+func valSetTotalVotingPower(valSet *ValidatorSet) int64 {
```

Comment:
we should not duplicate domain logic in tests

Reply:
In general yes. But I need the sum of voting power and cannot use the TotalVotingPower() method here.

Reply:
Can't you test it separately and here assume it's working properly.

```
set := construct a set with 33 tvp 
assert.Equal(t, 33, set.TotalVotingPower())
```

Reply:
I could but here we want to make sure that the totalVotingPower is updated as part of the updateWithChangeSet(). So even if the TotalVotingPower() works correctly it doesn't help if we don't call it.
One possibility is to make a private method in validator_set.go to compute totalVotingPower all the time and have TotalVotingPower() and UpdateWithChangeSet() call that instead. I can then use this private method in testing.

Reply:
😄 therefore we've arrived to where we started because if you "use this private method in testing." you are not testing anything.

> I could but here we want to make sure that the totalVotingPower is updated as part of the updateWithChangeSet()

Can we compare two values? old != new

Reply:
I am testing that vals.totalVotingPower value is correct. I am not interested to test that the private method is correct, just that it has been called.

Reply:
> Can we compare two values? old != new
Don't have access to old here. This function checks that after processing, the validator set has some properties:
- totalVotingPower is computed
- is scaled
- is centered
Note also how I call computeMaxMinPriorityDiff(), another private method to check the centering.

Reply:
> * totalVotingPower is computed

IMHO, there could be 2 ways to approach this: either make sure `valSet.totalVotingPower != 0` (this wouldn't be the case if `TotalVotingPower()` was not called; at leaset for freshly created validator sets), or, save away `valSet.totalVotingPower` in a variable and call `TotalVotingPower()` after: both values should be the same (and the latter also computes the sum you need without replicating the logic in a test).